File: ios/Tehilim/Core/Search/SearchInterpreter.swift
import Foundation

/// Une occurrence trouvée dans le texte d'un verset (V2.3 — recherche plein-texte).
struct VerseTextMatch: Identifiable {
    let psalm: Psalm
    let verse: Verse
    let snippet: String          // extrait à afficher (hébreu ou traduction)
    var id: String { verse.id }
}

struct SearchQueryResult {
    let exactMatch: Psalm?
    let suggestions: [Psalm]
    let interpretation: String?
    /// V2.3 — occurrences du terme dans le texte des Tehilim (hébreu + traduction).
    var textMatches: [VerseTextMatch] = []
}

struct SearchInterpreter {
    let repository: PsalmRepository

    private static let stripWords: Set<String> = [
        "tehilim", "tehillim", "psaume", "psaumes", "psalm", "תהילים", "תהלים"
    ]

    /// Plafond d'occurrences plein-texte (perf + lisibilité de la liste).
    private static let maxTextMatches = 60

    func interpret(_ rawQuery: String) -> SearchQueryResult {
        let trimmed = rawQuery.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !trimmed.isEmpty else {
            return .init(exactMatch: nil, suggestions: defaultSuggestions(), interpretation: nil)
        }

        // 1. Retirer les mots prefix
        var cleaned = trimmed
        for word in Self.stripWords {
            cleaned = cleaned.replacingOccurrences(of: word, with: " ")
        }
        cleaned = cleaned
            .replacingOccurrences(of: "\"", with: "")
            .replacingOccurrences(of: "'", with: "")
            .replacingOccurrences(of: "״", with: "")
            .replacingOccurrences(of: "׳", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        // 2. Tentative numéro arabe
        if let arabicNumber = extractArabicNumber(cleaned), let psalm = repository.psalm(id: arabicNumber) {
            return .init(
                exactMatch: psalm,
                suggestions: nearby(id: psalm.id),
                interpretation: "Tehilim \(psalm.id) · \(psalm.hebrewNumber)"
            )
        }

        // 3. Tentative hébraïque (numéro en lettres)
        if let hebrewNumber = HebrewNumerals.toInt(cleaned), let psalm = repository.psalm(id: hebrewNumber) {
            return .init(
                exactMatch: psalm,
                suggestions: nearby(id: psalm.id),
                interpretation: "Tehilim \(psalm.id) · \(psalm.hebrewNumber)"
            )
        }

        // 4. Recherche plein-texte (V2.3) — dès 2 caractères, sur le texte hébreu
        //    (sans nikud/téamim) et la traduction de la langue active.
        let textMatches = trimmed.count >= 2 ? searchText(rawQuery) : []
        return .init(exactMatch: nil, suggestions: textMatches.isEmpty ? defaultSuggestions() : [],
                     interpretation: nil, textMatches: textMatches)
    }

    // MARK: - Recherche plein-texte

    private func searchText(_ rawQuery: String) -> [VerseTextMatch] {
        // Normalisation hébraïque : on retire nikud + téamim des deux côtés
        // pour une recherche tolérante (l'utilisateur tape rarement la vocalisation).
        let hebQuery = Self.stripHebrewDiacritics(rawQuery)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        // Normalisation latine : minuscules + suppression des accents.
        let latinQuery = rawQuery.folding(options: [.diacriticInsensitive, .caseInsensitive],
                                          locale: .current)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let hasHebrew = hebQuery.unicodeScalars.contains { (0x05D0...0x05EA).contains($0.value) }
        let lang = AppLocale.translationLanguage

        var matches: [VerseTextMatch] = []
        for psalm in repository.allPsalms {
            for verse in psalm.verses {
                var snippet: String?
                if hasHebrew, !hebQuery.isEmpty {
                    let stripped = Self.stripHebrewDiacritics(verse.hebrew)
                    if stripped.contains(hebQuery) { snippet = verse.hebrew }
                }
                if snippet == nil, !hasHebrew, latinQuery.count >= 2,
                   let t = verse.translation(for: lang) {
                    let folded = t.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
                    if folded.contains(latinQuery) { snippet = t }
                }
                if let snippet {
                    matches.append(VerseTextMatch(psalm: psalm, verse: verse,
                                                  snippet: Self.trim(snippet)))
                    if matches.count >= Self.maxTextMatches { return matches }
                }
            }
        }
        return matches
    }

    /// Retire nikud (U+05B0–U+05BD, U+05BF, U+05C1–U+05C2, U+05C4–U+05C5, U+05C7)
    /// et téamim (U+0591–U+05AF) — tout le bloc de marques hébraïques.
    static func stripHebrewDiacritics(_ s: String) -> String {
        String(String.UnicodeScalarView(s.unicodeScalars.filter { scalar in
            !(0x0591...0x05CF).contains(scalar.value)
        }))
    }

    private static func trim(_ s: String, limit: Int = 90) -> String {
        s.count <= limit ? s : String(s.prefix(limit)) + "…"
    }

    // MARK: - Helpers

    private func extractArabicNumber(_ s: String) -> Int? {
        let digits = s.filter { $0.isNumber }
        guard !digits.isEmpty, let n = Int(digits), (1...150).contains(n) else { return nil }
        return n
    }

    private func nearby(id: Int) -> [Psalm] {
        let candidates = [id - 1, id + 1].compactMap { repository.psalm(id: $0) }
        return candidates
    }

    private func defaultSuggestions() -> [Psalm] {
        [23, 27, 91, 121, 130, 150].compactMap { repository.psalm(id: $0) }
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/SearchInterpreter.kt
package com.david.tehilim.core.service

import com.david.tehilim.core.model.Psalm
import com.david.tehilim.core.model.TranslationLanguage
import com.david.tehilim.core.model.Verse

/** Occurrence d'un terme dans le texte d'un verset (V2.3 — recherche plein-texte). */
data class VerseTextMatch(
    val psalm: Psalm,
    val verse: Verse,
    val snippet: String
)

/**
 * Résultat structuré d'une recherche Tehilim.
 *
 * - [exactMatch] : Tehilim trouvé par numéro arabe ou gematria hébraïque.
 * - [suggestions] : Tehilim voisins (id-1, id+1) si match, sinon liste par défaut.
 * - [interpretation] : libellé court "Tehilim N · hebrewNumber" pour feedback UI.
 */
data class SearchQueryResult(
    val exactMatch: Psalm?,
    val suggestions: List<Psalm>,
    val interpretation: String?,
    /** V2.3 — occurrences du terme dans le texte (hébreu + traduction). */
    val textMatches: List<VerseTextMatch> = emptyList()
)

/**
 * Interprète une requête utilisateur ("23", "psaume 23", "כג", "tehilim 91"…).
 * Mirror direct de `SearchInterpreter.swift` iOS.
 */
class SearchInterpreter(private val repository: PsalmRepository) {

    companion object {
        private val STRIP_WORDS = setOf(
            "tehilim", "tehillim", "psaume", "psaumes", "psalm", "תהילים", "תהלים"
        )

        private val DEFAULT_SUGGESTION_IDS = listOf(23, 27, 91, 121, 130, 150)
        private const val MAX_TEXT_MATCHES = 60
    }

    fun interpret(
        rawQuery: String,
        translationLanguage: TranslationLanguage = TranslationLanguage.EN
    ): SearchQueryResult {
        val trimmed = rawQuery.trim().lowercase()
        if (trimmed.isEmpty()) {
            return SearchQueryResult(
                exactMatch = null,
                suggestions = defaultSuggestions(),
                interpretation = null
            )
        }

        // 1. Retirer les mots prefix ("psaume 23" → "23")
        var cleaned = trimmed
        for (word in STRIP_WORDS) {
            cleaned = cleaned.replace(word, " ")
        }
        cleaned = cleaned
            .replace("\"", "")
            .replace("'", "")
            .replace("״", "")
            .replace("׳", "")
            .trim()

        // 2. Numéro arabe
        extractArabicNumber(cleaned)?.let { n ->
            repository.psalm(n)?.let { psalm ->
                return SearchQueryResult(
                    exactMatch = psalm,
                    suggestions = nearby(psalm.id),
                    interpretation = "Tehilim ${psalm.id} · ${psalm.hebrewNumber}"
                )
            }
        }

        // 3. Numéro hébraïque (gematria)
        HebrewNumerals.toInt(cleaned)?.let { n ->
            repository.psalm(n)?.let { psalm ->
                return SearchQueryResult(
                    exactMatch = psalm,
                    suggestions = nearby(psalm.id),
                    interpretation = "Tehilim ${psalm.id} · ${psalm.hebrewNumber}"
                )
            }
        }

        // 4. Recherche plein-texte (V2.3) — dès 2 caractères.
        val textMatches = if (trimmed.length >= 2) searchText(rawQuery, translationLanguage) else emptyList()
        return SearchQueryResult(
            exactMatch = null,
            suggestions = if (textMatches.isEmpty()) defaultSuggestions() else emptyList(),
            interpretation = null,
            textMatches = textMatches
        )
    }

    private fun searchText(rawQuery: String, lang: TranslationLanguage): List<VerseTextMatch> {
        val hebQuery = stripHebrewDiacritics(rawQuery).trim()
        val latinQuery = foldLatin(rawQuery).trim()
        val hasHebrew = hebQuery.any { it.code in 0x05D0..0x05EA }
        val out = ArrayList<VerseTextMatch>()
        for (psalm in repository.allPsalms) {
            for (verse in psalm.verses) {
                var snippet: String? = null
                if (hasHebrew && hebQuery.isNotEmpty()) {
                    if (stripHebrewDiacritics(verse.hebrew).contains(hebQuery)) snippet = verse.hebrew
                }
                if (snippet == null && !hasHebrew && latinQuery.length >= 2) {
                    verse.translation(lang)?.let { t ->
                        if (foldLatin(t).contains(latinQuery)) snippet = t
                    }
                }
                if (snippet != null) {
                    out.add(VerseTextMatch(psalm, verse, trimSnippet(snippet!!)))
                    if (out.size >= MAX_TEXT_MATCHES) return out
                }
            }
        }
        return out
    }

    /** Retire nikud + téamim (bloc U+0591–U+05CF) pour une recherche tolérante. */
    private fun stripHebrewDiacritics(s: String): String =
        s.filter { it.code !in 0x0591..0x05CF }

    /** Minuscules + suppression des accents (recherche latine tolérante). */
    private fun foldLatin(s: String): String =
        java.text.Normalizer.normalize(s.lowercase(), java.text.Normalizer.Form.NFD)
            .replace(Regex("\\p{Mn}+"), "")

    private fun trimSnippet(s: String, limit: Int = 90): String =
        if (s.length <= limit) s else s.take(limit) + "…"

    private fun extractArabicNumber(s: String): Int? {
        val digits = s.filter { it.isDigit() }
        if (digits.isEmpty()) return null
        val n = digits.toIntOrNull() ?: return null
        return if (n in 1..150) n else null
    }

    private fun nearby(id: Int): List<Psalm> =
        listOfNotNull(repository.psalm(id - 1), repository.psalm(id + 1))

    private fun defaultSuggestions(): List<Psalm> =
        DEFAULT_SUGGESTION_IDS.mapNotNull { repository.psalm(it) }
}
